Add MCO Flattened Ignition proposal

hardys opened the pull request:
Draft proposal to add MCO management of a "flattened" ignition configuration - this could be used to break the chicken/egg problem evident in some network configurations (particularly in the case of baremetal), e.g https://bugzilla.redhat.com/show_bug.cgi?id=1824331

hardys commented:
/cc @celebdor @stbenjam @kirankt

celebdor commented on enhancements/machine-config/mco-flattened-ignition.md:
```diff
@@ -0,0 +1,176 @@
+  * How do we resolve the issue of CAPBM not being able to access the MCS,
+    can we reliably determine the MCO managed resource and reference it directly?
+  * Does this approach fit with discussions around declarative network
+    configuration (for day-1 and day-2)?
```
My take on this is that the declarative network configuration for day 1 whether it is written in a MachineConfig or another CR would be rendered to the flattened ignition in the same way as described in this proposal.

hardys commented on enhancements/machine-config/mco-flattened-ignition.md:
```diff
@@ -0,0 +1,176 @@
+For UPI deployments, it may be possible to work around this situation when
+using coreos-install, since this has a `--copy-network` option which can drop
+NetworkManager config into /boot, but it would be a more consistent user
+experience if all configuration was possible via ignition?
```
Yeah, I was leaving it as a question so those more involved with UPI can add their opinion - to me it makes a lot of sense to have a consistent interface but I didn't want to presume that as a goal without consensus :)

stbenjam commented on enhancements/machine-config/mco-flattened-ignition.md:
```diff
@@ -0,0 +1,176 @@
+The ironic terraform provider is designed to be generic, and thus we would prefer
+not to add OS specific handling of the user-data there, but it would be technically
+possible to do the flattening in some terraform provider, and duplicate that
+logic in the cluster-api-provider-baremetal.
```
I experimented with this a little bit using an ignition data source that can create the flattened ignition (https://github.com/openshift-metal3/terraform-provider-openshift).

The problem is that terraform is frustratingly limited when you consider non-cloud cases. I constantly run into issues related to wanting to use terraform to provision infrastructure you'll use later by other terraform resources.  This is one of them: it needs the ignition for the control plane prior to creating the bootstrap.

I had some crazy ideas about throwing the control plane terraform code away, running cluster-baremetal-operator on the bootstrap, and having it provision the control plane. But, it's not a fully thought through idea.

hardys replied:
@stbenjam thanks for clarifying, I'd forgotten that detail - would you like to propose some alternative text to explain why it's actually not possible, or shall I just rework this to say it's been investigated and isn't possible due to terraform limitations?

cgwalters replied:
(this is a tangent)

I can't speak for the installer team, but I think there's general agreement that Terraform is an implementation detail today - we aren't tied to it forever.  And certainly whenever we go to add a kubevirt backend to the installer (yes it makes sense for testing at least) using TF to write Kube objects will be...weird.

I think indeed it would be better if we moved to mostly having TF provision the bootstrap node and *some* associated cloud resources but I know the installer team was concerned about speed.  Not clear to me how valid that is.

hardys replied:
Ok I'll try to clarify this, I guess the main point is we don't want to do it this way due to duplication and terraform limitations, it will be preferable to have all the ignition artifacts managed in one place (the MCO)

hardys replied:
I added a comment that this isn't implementable based on previous experience prototyping

cgwalters commented on enhancements/machine-config/mco-flattened-ignition.md:
```diff
@@ -0,0 +1,176 @@
+
+### Convert IPI to coreos-install based deployment
+
+Long term, from an OpenShift perspective, this probably makes sense, however
```
The way I'd say it is that there's massive overlap between the two - they're doing the same thing really.

hardys replied:
Yeah agreed, the issue is the well-tested path for Ironic at this point is to use the ironic community supported agent/ramdisk (which is why I said from an OpenShift perspective, we want to support a single solution for OpenShift, but the Ironic community will still want to maintain the existing solution).

This is solveable but is likely to take us some time (maybe create an FCOS based upstream CI job, deal with interface differences, etc) so it's being discounted as a viable solution in the 4.7 timeframe.

hardys replied:
I added a comment to clarify that there is overlap, and we're agreed that long-term this makes sense.

cgwalters reviewed (commented):
Let's close https://github.com/openshift/machine-config-operator/issues/1690 in favor of this please.

Overall...I'm OK with this.  The text notes that long term we want to move baremetal IPI to the Live ISO which solves a whole lot of problems in this area.

cgwalters commented on enhancements/machine-config/mco-flattened-ignition.md:
```diff
@@ -0,0 +1,176 @@
+(where the native vlan is often used for PXE provisioning),
+with the controlplane network configured via a VLAN on the same physical interface.
+
+For UPI deployments, it may be possible to work around this situation when
```
I'd just link to https://github.com/openshift/enhancements/blob/master/enhancements/rhcos/static-networking-enhancements.md

(Which also leads to the point that the CoreOS team considers these problems solved via the Live ISO as noted below)

hardys replied:
Done

hardys commented:
> Let's close [openshift/machine-config-operator#1690](https://github.com/openshift/machine-config-operator/issues/1690) in favor of this please.

Now closed.

> Overall...I'm OK with this. The text notes that long term we want to move baremetal IPI to the Live ISO which solves a whole lot of problems in this area.

Thanks for the review!  Yes I added clarification that long term we want to address that overlap.

Sounds like we're OK to proceed with some implementation work on this pending further reviews, I'll take a look and get something up as a WIP PR.

There are two remaining open questions that will influence the implementation:

What kind of resource should we use to store the flattented config - I'm thinking a MachineConfig like the rendered config?

How do we access the flattened config in the machine-api actuator, given that the MCS isn't accessible from services running on the cluster?  I'm thinking we ensure it's possible to derive which config to use (either by role or some additional metadata) for any given machine, then access the resource directly via the kube API (will the machine controller API creds allow this, given that the namespaces are different?)

hardys commented:
/retitle Add MCO Flattened Ignition proposal

craychee commented:
@kikisdeliveryservice @yuqi-zhang @sinnykumari @ericavonb @runcom

runcom commented on enhancements/machine-config/mco-flattened-ignition.md:
```diff
@@ -0,0 +1,178 @@
+pointer ignition config, stored in a Secret, and the rendered config stored
+in a MachineConfig object (both per role, not per machine).
+
+Both artifacts are accessible via the MCS (although not currently from within
```
uhm, the pointer ignition config isn't "accessible" via the MCS so s/both artifacts/just the rendered config/g

hardys commented:
Updated to account for feedback so far (thanks!) - also added more design details based on initial investigations, and some details of CI coverage we can use to test this.

hardys commented on enhancements/machine-config/mco-flattened-ignition.md:
```diff
@@ -0,0 +1,191 @@
+
+### Version Skew Strategy
+
+## Alternatives
```
Another alternative could be to just merge the installer generated ignition config into the rendered config (I guess we'd need to create a MachineConfig object derived from the installer generated user-data secret, then rely on the existing MergeMachineConfigs logic in MCO.

@runcom that would also solve the issue with the management of the pointer ignition, since we'd convert any data from the installer-created ignition to a MachineConfig, thus the pointer config can just be statically tempated as in your recent implementation?

hardys replied:
To clarify this idea, something like:

- Installer creates master-user-data (as it does today, [may contain user ignition customizations](https://github.com/openshift/installer/blob/master/docs/user/customization.md#os-customization-unvalidated))
- MCO parses master-user-data and creates a MachineConfig with the ignition data
- MCO generates master-user-data-managed from template (as implemented in https://github.com/openshift/machine-config-operator/pull/1792 but reverted via https://github.com/openshift/machine-config-operator/pull/2126 due to breaking the installer interface)
- Any installer provided customizations show up in the existing rendered config (same as if the user had provided them via MachineConfig manifests)

We could potentially flag some warning when this happens, and use it as a migration path to eventually deprecate the direct ignition-configs customization in favor of MachineConfig manifests?

This would also solve the issue for IPI baremetal without any MCO API changes, since we could consume the existing rendered config directly (we did this previously ref https://github.com/openshift/installer/pull/3276 but that was also reverted for the same reason as above)

@cgwalters @crawford any thoughts?

hardys replied:
I guess there are some RBAC considerations:

- The current user-data secret ends up in the openshift-machine-api namespace - https://github.com/openshift/machine-config-operator/pull/1792 worked around that by generating the pointer ignition manifest in the right namespace
- We could adjust the installer to write future user-data to the openshift-machine-config-operator namespace (so it can be read and converted into a  MachineConfig resource)
- On upgrade the plan from https://github.com/openshift/machine-config-operator/pull/1792 remains, e.g just reference the existing non-managed user-data secret?

hardys replied:
I looked at the code, and it's probably simplest to have the installer generate a MachineConfig manifest that contains any config provided via `create ignition-configs`

That avoids any changes to the MCO (other than restoring the reverted pointer-ignition change from @runcom) and potentially allows us to warn the user when config is provided via that interface.

yuqi-zhang replied:
I think I'm generally in favour of the MCO managing the stub configs coming from installer in one way or another, and this approach will probably work. I think we should keep the following points in mind:
1. Since the stub configs were never changed before we never considered any dependency between the MAO and the MCO. This method would likely (as I understand it) introduce an ordering where the MCO has to process the stub config first before machines can be booted.
2. We implicitly supported "per-node" configuration with the stub config, for example after you generated the stub config, you could configure each node's networking separately before boot with a different static IP via the ignition stub, and the MCO would be fine with it since it had no understanding of what existed in the stub. I guess for this example, the generated configs with `openshift-installer create ignition-configs` would have been supplied to the nodes manually. Is this a case we care about?

runcom replied:
good points, I care about solving point number 2 as it seems we really never wanted to advertise the use of that (last we checked with @crawford )

hardys replied:
> I think I'm generally in favour of the MCO managing the stub configs coming from installer in one way or another, and this approach will probably work. I think we should keep the following points in mind:
> 
>     1. Since the stub configs were never changed before we never considered any dependency between the MAO and the MCO. This method would likely (as I understand it) introduce an ordering where the MCO has to process the stub config first before machines can be booted.

I think we can avoid this if instead the installer creates a MachineConfig manifest with any user customizations it finds on the `create cluster` phase, and we could either skip generating this or leave it empty in the case where the stub config output via the `create ignition-configs` phase is unmodified?

We can also warn the user if the decision is made to deprecate the modification of the stub config with this approach (which may be harder if we have the MCO process the stub config, in addition to the ordering concern you raise above).
 
>     2. We implicitly supported "per-node" configuration with the stub config, for example after you generated the stub config, you could configure each node's networking separately before boot with a different static IP via the ignition stub, and the MCO would be fine with it since it had no understanding of what existed in the stub. I guess for this example, the generated configs with `openshift-installer create ignition-configs` would have been supplied to the nodes manually. Is this a case we care about?

Good point - IIUC this has only ever worked for the UPI case, where you download the stub config from the installer `create ignition-configs` phase, then host some per-host modified configs somewhere outside of the cluster?

In the IPI case, there's only a single secret per role, so per-node config is not currently supported via that workflow.

So, I think to ensure both workflows continue to work the same as today, we just need to ensure that the new MachineConfig object either isn't generated or is empty at the `create ignition-configs` stage, and that we can detect and re-generate that asset in the case where some user-customization happened between `create ignition-configs` and `create cluster` (I guess we can compare the assets loaded from file with that generated inside the installer).

hardys commented:
Superseded by https://github.com/openshift/enhancements/pull/540
